srcjs/output_binding_image.js

```diff
@@ -240,162 +259,221 @@ imageutils.initPanelScales = function(coordmap) {
+  coordmap.scaleCssToImg = function(offset_css) {
+    const pixel_scaling = coordmap.imgToCssScalingRatio();
+
+    const result = mapValues(offset_css, (value, key) => {
+      const prefix = key.substring(0, 1);
+      
+      if (prefix === "x") {
+        return offset_css[key] / pixel_scaling.x;
+      } else if (prefix === "y") {
+        return offset_css[key] / pixel_scaling.y;
+      }
+      return null;
+    });
+
+    return result;
+  };
+
+  // Given an offset in an img, in source image pixels, return the
+  // corresponding offset in CSS pixels. If the img content is 1000 pixels
+  // wide, but is scaled to 400 pixels on screen, and the input is x:1000,
+  // then this will return x:400.
+  coordmap.scaleImgToCss = function(offset_img) {
+    const pixel_scaling = coordmap.imgToCssScalingRatio();
+
+    const result = mapValues(offset_img, (value, key) => {
+      const prefix = key.substring(0, 1);
+      
+      if (prefix === "x") {
+        return offset_img[key] * pixel_scaling.x;
+      } else if (prefix === "y") {
+        return offset_img[key] * pixel_scaling.y;
+      }
+      return null;
+    });
+
+    return result;
+  };
+
+  // Returns the x and y ratio the image content is scaled to on screen. If
+  // the image data is 1000 pixels wide and is scaled to 300 pixels on screen,
+  // then this returns 0.3. (Note the 300 pixels refers to CSS pixels.)
+  coordmap.imgToCssScalingRatio = function() {
+    const img_dims = findDims($img);
     return {
-      xmin: Math.min(offset1.x, offset2.x),
-      xmax: Math.max(offset1.x, offset2.x),
-      ymin: Math.min(offset1.y, offset2.y),
-      ymax: Math.max(offset1.y, offset2.y)
+      x: img_dims.x / coordmap.dims.width,
+      y: img_dims.y / coordmap.dims.height
     };
   };
 
+  coordmap.cssToImgScalingRatio = function() {
+    const res = coordmap.imgToCssScalingRatio();
+    return {
+      x: 1 / res.x,
+      y: 1 / res.y
+    };
+  };
 
-  // Shift an array of values so that they are within a min and max.
-  // The vals will be shifted so that they maintain the same spacing
-  // internally. If the range in vals is larger than the range of
-  // min and max, the result might not make sense.
-  coordmap.shiftToRange = function(vals, min, max) {
-    if (!(vals instanceof Array))
-      vals = [vals];
-
-    var maxval = Math.max.apply(null, vals);
-    var minval = Math.min.apply(null, vals);
-    var shiftAmount = 0;
-    if (maxval > max) {
-      shiftAmount = max - maxval;
-    } else if (minval < min) {
-      shiftAmount = min - minval;
-    }
+  // Given an offset in css pixels, return an object representing which panel
+  // it's in. The `expand` argument tells it to expand the panel area by that
+  // many pixels. It's possible for an offset to be within more than one
+  // panel, because of the `expand` value. If that's the case, find the
+  // nearest panel.
+  coordmap.getPanelCss = function(offset_css, expand = 0) {
```

Comment:
No need to gather up all the panel objects that match--you can instead just keep track of the nearest one seen so far.